File: runtime/python/lexerHelpers.py
from .errors import InputError

def contains(t, string, times): return t.val.count(string) == times
def isUpper(t, i): return t.val[i].isupper()
def fullyApplies(t, r): 
    try:
        _, rest = r(([], t.val))
        return rest == ""
    except InputError:
        return False
def applies(t, r): 
    try:
        r(([], t.val))
        return True
    except InputError:
        return False

File: runtime/dotnet/LexerHelpers.fs
module LexerHelpers
open Token
open ParserCombinator
open System.Text.RegularExpressions
open System

let contains (tok: TokenInterface) str times = Regex.Matches(tok.value, Regex.Escape str).Count = times
let isUpper (tok: TokenInterface) i = Char.IsUpper(tok.value.[i])
let fullyApplies (tok: TokenInterface) (rule: Value -> FunRes) = 
    match rule ([], tok.value) with
    | Ok((_, "")) -> true
    | Ok(_) -> false
    | Error(_) -> false
let applies (tok: TokenInterface) (rule: Value -> FunRes) = 
    match rule ([], tok.value) with
    | Ok(_) -> true
    | Error(_) -> false